### PR TITLE
Mqtt annotations

### LINKED ISSUE #304: MQTT Annotations
add annotations to subscribe a method to a topic with wildcards and receive the message and the topic

### LINKED ISSUE #412: Spotify mute online works after using spotify plugin once
Spotify mute only works after using the plugin at least once.

Possible solutions:
* convert `PlayerLogic` to a RunnableService and connect to Spotify API on startup.
* declarative MessageHub API that starts up `PlayerLogic` on demand (disadvantage: possible delay on first use)

### PR DESCRIPTION
**Pull request by Legion2**
added annotations for MessageReveiver with Subscriptions
implements deployment adapter
close #304 
fixed #412 

Pull request status:
- [x] Tests added/updated
- [ ] 80 % Coverage
- [x] Code conventions
- [x] [Doku](https://github.com/AmyAssist/Amy/wiki/Message-Hub#message-receiver) required and added
- [x] Ready for review

### CODE REVIEW

File: amy-message-hub/src/main/java/de/unistuttgart/iaas/amyassist/amy/messagehub/annotations/MessageRecievers.java [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+import de.unistuttgart.iaas.amyassist.amy.messagehub.topic.TopicFactory;
+import de.unistuttgart.iaas.amyassist.amy.messagehub.topic.TopicFilter;
+
+/**
+ * Loads the DD and subscribes all MessageReceivers.
+ * 
+ * @author Leon Kiefer
+ */
+@Service(MessageRecievers.class)
+public class MessageRecievers implements DeploymentContainerService {
```
Comment (Bensge) [inline amy-message-hub/src/main/java/de/unistuttgart/iaas/amyassist/amy/messagehub/annotations/MessageRecievers.java]: correct to MessageRec**ei**vers

File: amy-message-hub/src/main/java/de/unistuttgart/iaas/amyassist/amy/messagehub/internal/SubscriptionObject.java [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+
+		try {
+			if (this.withTopicName) {
+				this.method.invoke(createAndInitialize, msg.getPayload(), topic);
+			} else {
+				this.method.invoke(createAndInitialize, msg.getPayload());
+			}
+
+		} catch (IllegalAccessException e) {
+			throw new IllegalArgumentException("tryed to invoke method " + this.method + " but got an error", e);
```
Comment (Bensge) [inline amy-message-hub/src/main/java/de/unistuttgart/iaas/amyassist/amy/messagehub/internal/SubscriptionObject.java]: tr**ie**d

File: plugins/spotify/src/main/java/de/unistuttgart/iaas/amyassist/amy/plugin/spotify/logic/SpotifyMessageReciever.java [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+@MessageReceiver
+public class SpotifyMessageReciever {
+
+	private static final String MUSIC_MUTE_TOPIC_FUNCTION = "muteMusic";
+	@Reference
+	private PlayerLogic player;
+	@Reference
+	private Logger logger;
+	@Reference
+	private SpotifyAPICalls spotifyAPICalls;
```
Comment (Bensge) [inline plugins/spotify/src/main/java/de/unistuttgart/iaas/amyassist/amy/plugin/spotify/logic/SpotifyMessageReciever.java]: unused variable, consider removing it

File: plugins/spotify/src/main/java/de/unistuttgart/iaas/amyassist/amy/plugin/spotify/logic/SpotifyMessageReciever.java [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+
+	/**
+	 * 
+	 * @param message
+	 *            "true" or "false"
+	 */
+	@Subscription("smarthome/+/+/mute")
+	public void mute(String message) {
+		switch (message) {
```
Comment (Bensge) [inline plugins/spotify/src/main/java/de/unistuttgart/iaas/amyassist/amy/plugin/spotify/logic/SpotifyMessageReciever.java]: move this duplicated code into one internal method maybe?

File: plugins/spotify/src/main/java/de/unistuttgart/iaas/amyassist/amy/plugin/spotify/logic/SpotifyMessageReciever.java [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+		case "true":
+			this.player.setSuppressed(true);
+			break;
+		case "false":
+			this.player.setSuppressed(false);
+			break;
+		default:
+			this.logger.warn("unkown message {}", message);
```
Comment (Bensge) [inline plugins/spotify/src/main/java/de/unistuttgart/iaas/amyassist/amy/plugin/spotify/logic/SpotifyMessageReciever.java]: spelling: unk**n**own

**Comment by Bensge** [comment]
Spotify integrations works well 👍 
The first time it takes about two or three seconds to mute Spotify, but that's fine. After that, it works just like before (before this PR, it did not work the first time around)

File: amy-message-hub/src/main/java/de/unistuttgart/iaas/amyassist/amy/messagehub/internal/SubscriptionUtil.java [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+	 * 
+	 * @param method
+	 *            the method to validate as Subscription method
+	 * @throws IllegalArgumentException
+	 *             if the given method is not a valid Subscription method
+	 */
+	public static void assertValidSubscriptionMethod(Method method) {
+		if (!method.getReturnType().equals(Void.TYPE))
+			throw new IllegalArgumentException(
+					"The returntype of a method annotated with @Subscription should be String.");
```
Comment (neumantm) [inline amy-message-hub/src/main/java/de/unistuttgart/iaas/amyassist/amy/messagehub/internal/SubscriptionUtil.java]: Wrong: "should be Void."

File: amy-message-hub/src/main/java/de/unistuttgart/iaas/amyassist/amy/messagehub/internal/SubscriptionObject.java [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+			this.withTopicName = true;
+		} else {
+			throw new IllegalArgumentException(
+					"Subscription can only have the message and the topicname as parameter.");
+		}
+	}
+
+	@Override
+	public void handle(TopicName topic, Message msg, ServiceLocator serviceLocator) {
+		Object createAndInitialize = serviceLocator.createAndInitialize(this.messageReceiverClass);
```
Comment (neumantm) [inline amy-message-hub/src/main/java/de/unistuttgart/iaas/amyassist/amy/messagehub/internal/SubscriptionObject.java]: rename

**Pull request merged**